app/routes/app/rewards/addresses.tsx

```diff
@@ -171,42 +167,83 @@ function AddressCards({ wallets }: { wallets: any }) {
   );
 }
 
-const schema = z.object({
+const addWalletSchema = z.object({
   payment: z.discriminatedUnion("tokenSymbol", [
     z.object({ tokenSymbol: z.literal("ETH"), address: EthAddressSchema }),
     z.object({ tokenSymbol: z.literal("SOL"), address: SolAddressSchema }),
     z.object({ tokenSymbol: z.literal("MATIC"), address: EthAddressSchema }),
   ]),
+  userId: z.string({ description: "The ID of the user the wallet belongs to." }),
+});
+
+const updateWalletSchema = z.object({
+  userId: z.string({ description: "The ID of the user the wallet belongs to." }),
+  currentAddress: z.string({ description: "The current address of the wallet." }),
+  newAddress: z.string({ description: "The new address of the wallet." }),
+});
+
+const deleteWalletSchema = z.object({
+  currentAddress: z.string({ description: "The current address of the wallet." }),
 });
 
-const validator = withZod(schema);
+export const addWalletValidator = withZod(addWalletSchema);
+export const deleteWalletValidator = withZod(deleteWalletSchema);
+export const updateWalletValidator = withZod(updateWalletSchema);
+
 function AddAddressButton() {
   const { tokens } = useTypedLoaderData<typeof loader>();
   const [openedAdd, setOpenedAdd] = useState(false);
 
+  const { user } = useTypedLoaderData<typeof loader>();
+  const [selectedNetwork, setSelectedNetwork] = useState<string>("");
+  const [selectedAddress, setSelectedAddress] = useState<string>("");
+  const [validAddress, setValidAddress] = useState(false);
+
+  useEffect(() => {
+    if (selectedAddress && selectedNetwork) {
+      const isValid = validateAddress(selectedAddress, getNetworkByTokenSymbol(selectedNetwork));
+      setValidAddress(isValid);
+    }
+  }, [selectedAddress, selectedNetwork]);
+
   return (
     <>
       <Button className="mx-auto" onClick={() => setOpenedAdd(true)}>
         Add Address
       </Button>
       <Modal isOpen={openedAdd} onClose={() => setOpenedAdd(false)} title="Add an address">
-        <ValidatedForm validator={validator} className="space-y-5 mt-5">
+        <ValidatedForm
+          action="/app/rewards/wallet/add"
+          method="post"
+          validator={addWalletValidator}
+          className="space-y-5 mt-5"
+        >
           <div className="pb-44 pt-8">
-            <AddPaymentAddressForm tokens={tokens} />
+            <AddPaymentAddressForm
+              setSelectedNetwork={setSelectedNetwork}
+              setSelectedAddress={setSelectedAddress}
+              tokens={tokens}
+            />
+          </div>
+          <div className="invisible h-0 w-0">
+            <ValidatedInput type="hidden" id="userId" name="userId" value={user ? user : ""} />
```

Comment:
This inspired me to use a hidden input for user on my PR. I also used Error to show some text that the user must be signed in. Check it out:
https://github.com/MetricsDAO/xyz/pull/219/commits/f816a3c07205d7230590c61f8de6d798c79362e9

Reply:
ohh nice love it, will add although it shouldn't ever get to that state since we gate rewards on user. Yes hidden inputs are so clutch

Reply:
juussst kidddding 😊  We should be adding the user data on the backend via session and not trust the user id sent as an input from the client-side.

Reply:
hahahha you're right. Shit, thought I was being crafty and didn't even think about it